lib/rules/no-inner-declarations.js

```diff
@@ -34,53 +37,48 @@ module.exports = {
     create(context) {
 
         /**
-         * Find the nearest Program or Function ancestor node.
-         * @returns {Object} Ancestor's type and distance from node.
+         * Ensure that a given node is at a program or function body's root.
+         * @param {ASTNode} node Declaration node to check.
+         * @returns {void}
          */
-        function nearestBody() {
+        function check(node) {
             const ancestors = context.getAncestors();
-            let ancestor = ancestors.pop(),
-                generation = 1;
+            let ancestor = ancestors.pop();
 
-            while (ancestor && ["Program", "FunctionDeclaration",
-                "FunctionExpression", "ArrowFunctionExpression"
-            ].indexOf(ancestor.type) < 0) {
-                generation += 1;
-                ancestor = ancestors.pop();
+            if (
+                ancestor.type === "BlockStatement" &&
+            !!~validBlockStatementParent.indexOf(ancestor.parent.type)
+            ) {
+                return;
             }
 
-            return {
+            if ((~validParent.indexOf(ancestor.type))) {
+                return;
+            }
 
-                // Type of containing ancestor
-                type: ancestor.type,
+            let parentsListToCheck = ancestor.type === "BlockStatement" ? validBlockStatementParent : validParent;
 
-                // Separation between ancestor and node
-                distance: generation
-            };
-        }
-
-        /**
-         * Ensure that a given node is at a program or function body's root.
-         * @param {ASTNode} node Declaration node to check.
-         * @returns {void}
-         */
-        function check(node) {
-            const body = nearestBody(),
-                valid = ((body.type === "Program" && body.distance === 1) ||
-                    body.distance === 2);
-
-            if (!valid) {
-                context.report({
-                    node,
-                    messageId: "moveDeclToRoot",
-                    data: {
-                        type: (node.type === "FunctionDeclaration" ? "function" : "variable"),
-                        body: (body.type === "Program" ? "program" : "function body")
-                    }
-                });
+            while (ancestor && !~parentsListToCheck.indexOf(ancestor.type)) {
+                ancestor = ancestors.pop();
+                if (ancestor.type === "BlockStatement") {
+                    ancestor = ancestor.parent;
+                    parentsListToCheck = validBlockStatementParent;
+                } else {
+                    parentsListToCheck = validParent;
+                }
```

Comment:
Can we simplify this, I think the goal is just to find the lowest ancestor that is either `Program`, `FunctionDeclaration`, `FunctionExpression` or `ArrowFunctionExpression`. This seems unrelated to `validParent` and `validBlockStatementParent`.

Reply:
We could maybe use `astUtils.getUpperFunction`.

Reply:
> We could maybe use `astUtils.getUpperFunction`.

didn't know about this, I will take a look 👍

Reply:
done